File: Partition Equal Subset Sum[15-September-2023]/Solution.java
//  >> JAVA CODE <<

class Solution{
    static boolean check(int i,int sm,long target,int[] arr,int N,Boolean[][] dp)
    {
        if(sm==target) return true;
        if(i==N) return false;
        
        if(dp[i][sm]!=null) return dp[i][sm];
        
        if(sm+arr[i]>target)
            return dp[i][sm]=check(i+1,sm,target,arr,N,dp);
            
        return dp[i][sm]=check(i+1,sm,target,arr,N,dp) || check(i+1,sm+arr[i],target,arr,N,dp);
    }
    static int equalPartition(int N, int arr[])
    {
        long sum=0;
        
        for(int i=0;i<N;i++) sum+=arr[i];
        
        if(sum%2==1) return 0;
        
        long half=sum/2;
        
        Boolean[][] dp=new Boolean[N][(int)(sum)];
        
        return check(0,0,half,arr,N,dp)?1:0;
    }
}

File: Partition Equal Subset Sum[15-September-2023]/Solution.py
//   >> PYTHON CODE <<

class Solution:
    def check(self,i,sm,target,arr,N,dp):
        if(sm==target): return True
        if(i==N): return False
        
        if(dp[i][sm] is not None): return dp[i][sm]
        
        if(sm+arr[i]>target):
            dp[i][sm]=self.check(i+1,sm,target,arr,N,dp)
            return dp[i][sm]
        
        dp[i][sm]=self.check(i+1,sm+arr[i],target,arr,N,dp) or self.check(i+1,sm,target,arr,N,dp)
        return dp[i][sm]
        
    def equalPartition(self, N, arr):
        
        sm=sum(arr)
        if(sm%2==1): return False
        
        dp=[[None for i in range(sm)] for _ in range(N)]
        
        half=sm/2
        
        return self.check(0,0,half,arr,N,dp)